File: cs97si/ass-6/1308 Is It A Tree.py
import collections

def create_dic_graph(edges):
    graph = collections.defaultdict(list)
    for e in edges:
        graph[e[0]].append(e[1])
    return graph

def ans(edges):
    graph = create_dic_graph(edges)
    keys = graph.keys()

    dic = collections.defaultdict(int)
    for u in keys:
        if dic.has_key(u) == False:
            dic[u] = 0
        for v in graph[u]:
            dic[v] += 1 # in-degree

    root = 0
    cnt = 0
    for key in dic.keys():
        if dic[key] == 0:
            root = key
            cnt += 1
    if cnt > 1:
        #print "multiple roots"
        return False

    cnt = 0
    for key in dic.keys():
        if dic[key] > 1:
            #print "mutiple in-degree"
            cnt += 1

    if cnt > 0:
        return False
    return True


def add_edges(edges, line):
    for i in range(0, len(line), 2):
        edges.append([line[i], line[i + 1]])
    

def find_root(graph):
    keys = graph.keys()

    dic = collections.defaultdict(int)
    for u in keys:
        if u not in dic:
            dic[u] = 0
        for v in graph[u]:
            dic[v] += 1 # in-degree
    for key in dic.keys():
        if dic[key] == 0:
            return (key, dic)
    return (-1, dic)

def bfs(graph, root, nodes):
    q = [root]
    hist = {}
    while len(q) != 0:
        u = q.pop(0)
        if u in hist:
            return False
        hist[u] = True
        for v in graph[u]:
            q.append(v)
    
    for v in nodes.keys():
        if v not in hist:
            return False
    return True

def ans1(edges):
    '''
    bfs
    '''
    if len(edges) == 0:
        return True
    graph = create_dic_graph(edges)
    root, dic = find_root(graph) 
    if root == -1:
        return False
    return bfs(graph, root, dic)

def recur_util(graph, u, hist):
    if u in hist:
        return False

    hist[u] = True
    for v in graph[u]:
        ret = recur_util(graph, v, hist)
        if ret == False:
            return False
    return True

def dfs(graph, root, dic):
    hist = {}
    ret = recur_util(graph, root, hist)
    if ret == False:
        return False
    
    for v in dic.keys():
        if v not in hist:
            return False
    return True

def ans2(edges):
    '''
    adjancency list + dfs
    '''
    if len(edges) == 0:
        return True
    graph = create_dic_graph(edges)
    root, dic = find_root(graph) 
    if root == -1:
        return False
    return dfs(graph, root, dic)

cnt = 1
edges = []
while True:
    line = list(map(int, input().strip().split()))
    if len(line) == 0:
        continue
    
    if line[0] == -1:
        break

    add_edges(edges, line)
    if line[-1] == 0:
        edges.pop()
        if ans2(edges) == True:
            print ("Case %d is a tree."%(cnt))
        else:
            print ("Case %d is not a tree."%(cnt))
        edges = []
        cnt += 1

File: cs97si/ass-6/1308 Is It A Tree.cpp
#include <cstdio>
#include <map>
#include <vector>
#include <queue>
#include <utility>
using namespace std;

bool ans( map<int, vector<int> > graph)
{
	map<int, int> in_deg;
	for (map<int, vector<int> >::iterator it = graph.begin(); it != graph.end(); it++)
	{
		int u = it->first;
		if (in_deg.count(u) == 0)
			in_deg[u] = 0;
		vector<int> &vs = graph[u];
		for (int i = 0; i < vs.size(); i++)
		{
			int v = vs[i];
			if(in_deg.count(v) == 0)
				in_deg[v] = 0;
			in_deg[v]++;
		}
		
	}
	int root = 0;
	int cnt = 0;
	map<int, int>::iterator it;
	it = in_deg.begin();
	for(it; it != in_deg.end(); it++)
	{
		if(in_deg[it->first] == 0)
		{
			root = it->first;
			cnt++;
		}
	}
	if(cnt > 1)
	{
		return false;
	}

	cnt = 0;
	it = in_deg.begin();
	for(it; it != in_deg.end(); it++)
	{
		if(in_deg[it->first] > 1)
		{
			cnt++;
		}
	}
	if(cnt > 0)
	{
		return false;
	}

	queue<int> q;
	map<int, bool> vis;
	q.push(root);
	while (!q.empty()) {
		int u = q.front();
		q.pop();
		if(vis[u] == true)
		{
			return false;
		}
		vis[u] = true;
		vector<int> &vs = graph[u];
		for (int i = 0; i < vs.size(); i++) {
			int v = vs[i];
			q.push(v);
		}
	}
	for (map<int, bool>::iterator it = vis.begin(); it != vis.end(); it++)
		if (!it->second) 
			return false;
		
	it = in_deg.begin();
	for(it; it != in_deg.end(); it++)
	{
		if (!vis[it->first]) 
			return false;
	}
	
	return true;
}


int main() {
	int u, v, t = 0;
	scanf("%d %d", &u, &v);
	while (u >= 0) {
		t++;
		map<int, vector<int> > graph;
		while (u > 0) {
			//if (deg.count(u) == 0) deg[u] = 0;
			//if (deg.count(v) == 0) deg[v] = 0;
			graph[u].push_back(v);
			scanf("%d %d", &u, &v);
		}
		bool ok = ans(graph);
		/*
		bool ok = true;
		int root = 0;
		for (map<int, int>::iterator it = deg.begin(); it != deg.end(); it++)
			if (it->second > 1) {
				ok = false;
				break;
			}
			else if (it->second == 0) {
				if (root > 0) {
					ok = false;
					break;
				}
				else root = it->first;
			}
		if (ok) {
		}
		*/
		if (ok) 
			printf("Case %d is a tree.\n", t);
		else 
			printf("Case %d is not a tree.\n", t);
		scanf("%d %d", &u, &v);
	}
}
